File: Code/backend/app/services/crs/crs_types.py
"""CRS 核心类型定义：分类枚举、CRSDef 数据类、CoordinatePoint。

设计要点：
- ``CRSCategory`` 继承 ``str, Enum`` 以便 JSON 序列化为纯字符串
- ``CRSDef`` 使用 ``frozen=True`` 保证注册项可安全共享
- ``CoordinatePoint`` 字段与旧 ``coordinate_transform_service.CoordinatePoint``
  完全一致（lng/lat），保证二进制兼容的 re-export
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class CRSCategory(str, Enum):
    """CRS 大类，决定转换路径。"""

    GEOGRAPHIC = "geographic"
    """地理坐标系（经纬度），如 EPSG:4326/4490/4258。"""

    PROJECTED = "projected"
    """投影坐标系（米/英尺），如 EPSG:3857/6933/UTM/Gauss-Krüger。"""

    ENCRYPTED = "encrypted"
    """加密坐标系（非标准 EPSG），如 GCJ-02/BD-09。仅中国境内使用。"""


@dataclass(frozen=True)
class CRSDef:
    """单个 CRS 定义。

    Attributes:
        code: 规范标识符。EPSG 系列为 ``'EPSG:XXXX'``；加密系为 ``'GCJ02'``/``'BD09'``（无前缀）。
        label: 中文显示名（用于前端下拉）。
        category: CRS 大类。
        epsg: 数值 EPSG 代码；加密系为 ``None``。
        proj4_def: proj4 定义字符串；加密系为 ``None``（由 ``_gcj_bd`` 模块处理）。
        area: 常用区域（如 ``'China'``/``'Europe'``/``'Global'``）。
        deprecated: 是否计划在 Phase 3+ 弃用。
    """

    code: str
    label: str
    category: CRSCategory
    epsg: int | None
    proj4_def: str | None
    area: str
    deprecated: bool = False


@dataclass(frozen=True)
class CoordinatePoint:
    """不可变 (lng, lat) 对。

    与旧 ``coordinate_transform_service.CoordinatePoint`` 字段一致，
    保证 deprecated 垫片的 re-export 二进制兼容。
    """

    lng: float
    lat: float

File: Code/frontend/src/services/crs/crs-types.ts
/**
 * CRS 模块类型定义（镜像后端 `crs_types.py`）。
 *
 * 字段命名采用 camelCase（前端惯例），与后端 snake_case 通过 `toApiPayload`
 * 序列化层转换。
 */

export type CRSCategory = 'geographic' | 'encrypted' | 'projected'

export interface CRSDef {
  /** CRS code，如 `'EPSG:4326'` / `'GCJ02'` / `'BD09'` */
  code: string
  /** 人类可读标签 */
  label: string
  category: CRSCategory
  /** EPSG 整数代码；加密系为 null */
  epsg: number | null
  /** proj4 定义串；加密系为 null */
  proj4Def: string | null
  /** 适用区域：`'Global'` / `'China'` / `'Europe'` */
  area: string
  /** 是否已废弃（Phase 2 保留 deprecated 垫片时为 true） */
  deprecated: boolean
}

/** 前端下拉用平铺项（`toApiPayload()` 返回） */
export interface CRSOption {
  code: string
  label: string
  category: CRSCategory
  area: string
  deprecated: boolean
}

/** 坐标点（与后端 `CoordinatePoint` 字段顺序一致：lng 在前，lat 在后） */
export interface CoordinatePoint {
  lng: number
  lat: number
}

/** 转换选项：偏移在 CRS 转换**后**应用（与后端一致） */
export interface TransformOptions {
  lngOffset?: number
  latOffset?: number
}

/** CRS 自动检测结果（镜像后端 `CRSDetectionResult`） */
export interface CRSDetectionResult {
  sourceCrs: string
  /** 0~1，反映检测来源可靠程度 */
  confidence: number
  method: 'rasterio_crs' | 'geojson_crs' | 'bounds_heuristic' | 'default'
  suggestedCrs: string
  needsUserConfirm: boolean
  /** 检测过程说明（前端展示或日志） */
  notes: string
}
